keps/sig-node/2400-node-swap/README.md

```diff
@@ -0,0 +1,847 @@
+behaviour under memory pressure, and applications cannot directly control what
+portions of their memory usage are swapped out. Since enabling swap permits
+greater memory usage for workloads in Kubernetes that cannot be predictably
+accounted for, it also increases the risk of noisy neighbours and unexpected
+packing configurations, as the scheduler cannot account for swap memory usage.
+
+This risk is mitigated by preventing any workloads from using swap by default,
+even if swap is enabled and available on a system. This will allow a cluster
+administrator to test swap utilization just at the system level without
+introducing unpredictability to workload resource utilization.
+
+Additionally, we will mitigate this risk by determining a set of metrics to
+quantify system stability and then gathering test and production data to
+determine if system stability changes when swap is available to the system
+and/or workloads in a number of different scenarios.
+
+Since swap provisioning is out of scope of this proposal, this enhancement
+poses low risk to Kubernetes clusters that will not enable swap.
+
+## Design Details
+
+We summarize the implementation plan as following:
+
+1. Add a feature gate `NodeSwapEnabled` to enable swap support.
+1. Leave the default value of kubelet flag `--fail-on-swap` to `true`, to avoid
+   changing default behaviour.
+1. Introduce a new kubelet config parameter, `MemorySwapLimit`.
+1. Introduce a new CRI parameter, `memory_swap_limit_in_bytes`.
+1. Integrate new kubelet config and pass values to CRI for container creation.
+1. Ensure container runtimes are updated so they can make use of the new CRI configuration.
+
+### Enabling swap as an end user
+
+Swap can be enabled as follows:
+
+1. Provision swap on the target worker nodes,
+1. Enable `NodeMemorySwap` flag on the kubelet,
+1. Set `--fail-on-swap` flag to `false`, and
+1. (Optional) Configure `MemorySwapLimit` in the KubeletConfig for tuning.
+
+### API Changes
+
+#### KubeConfig addition
+
+We will add an optional `MemorySwapLimit` value to the `KubeletConfig` struct
+in [pkg/kubelet/apis/config/types.go] for a compatible API change as follows:
+
+[pkg/kubelet/apis/config/types.go]: https://github.com/kubernetes/kubernetes/blob/6baad0a1d45435ff5844061aebab624c89d698f8/pkg/kubelet/apis/config/types.go#L81
+
+```go
+// KubeletConfiguration contains the configuration for the Kubelet
+type KubeletConfiguration struct {
+	metav1.TypeMeta
+...
+	// Configure swap memory available to container workloads.
+	// +featureGate=NodeSwapEnabled
+	// +optional
+	MemorySwap MemorySwapConfiguration
+}
+
+type MemorySwapConfiguration struct {
+	// Configure swap memory available to container workloads. May be one of
+	// "", "NoSwap": workloads cannot use swap
+	// "UnlimitedSwap": workloads can use unlimited swap, up to the allocatable limit.
+	// "LimitedSwap": workloads can use up to this limit of swap.
+	SwapBehavior string
+
+	LimitedSwap *LimitedSwapConfiguration
```

Comment:
maybe move this into a beta criteria.
this will get tricky with cgroups v2 and memory.high settings.